### ISSUE TITLE
appimagetool "doesn't look like a squashfs image."

### ISSUE THREAD

**Issue by probonopd** [probonopd/AppImageKit#282]
In the build from 36c4881 we get:

```
+chmod a+x ./appimagetool
+set +x
This doesn't look like a squashfs image.
Cannot mount AppImage, please check your FUSE setup.
You might still be able to extract the contents of this AppImage 
if you run it with the --appimage-extract option. 
See https://github.com/probonopd/AppImageKit/wiki/FUSE 
for more information
open dir error: No such file or directory
```

As a result, downstream users are affected:
* https://github.com/mypaint/mypaint/issues/742
* https://github.com/probonopd/AppImages/issues/136

 We should, in addition to the continuous builds, offer "last known good" builds.

To prevent this from happening again, we should "eat our own dogfood" and use appimagetool.AppImage in the CI process to build e.g, appimaged.AppImage; so that errors like this would block the Travis CI build job and not go unnoticed.

**Comment by probonopd** [comment]
Seems to have been caused by https://github.com/probonopd/AppImageKit/pull/278

**Comment by probonopd** [comment]
36c4881 also seems to have to do with it. Also reverted for now.

**Comment by mkrautz** [comment]
I'm confused. Does CI test i386? The code path I changed in elf.c is only ever executed against 32-bit ELFs, right?

**Comment by mkrautz** [comment]
However, I never did get appimagetool working properly on my CentOS 5 i386 environment (I got error messages like this one).

But I was using appimagetoolnoglib, and I attributed it to that. Manually assembling an AppImage by bundling the runtime+squashfs via cat did work for me.

**Comment by mkrautz** [comment]
The difference between my own assembled AppImage and the one generated by appimagetoolnoglib is that SquashFS starts at 0x20000:

`00020000  68 73 71 73 1e 00 00 00  49 46 2a 58 00 00 02 00  |hsqs....IF*X....|`

**Comment by mkrautz** [comment]
fs_offset is set to ...

`offset=46760`

although squashfs image starts at 0x20000...

**Comment by mkrautz** [comment]
Seems to me like the problem is that runtime.c doesn't consider an offset squashfs image? I.e. it wasn't taught about https://github.com/probonopd/AppImageKit/commit/29decf8db2d53c83c43b8027187bf00f9efa67b8?

**Comment by mkrautz** [comment]
Hmmm. Nevermind. Maybe it's that appimagetoolnoglib wasn't updated to support offsets instead? :-) (As I first thought.)

**Comment by mkrautz** [comment]
FWIW,

I can build an run AppImages just fine built with appimagetoolnoglib at 

with this patch applied:

	$ git diff appimagetoolnoglib.c 
	diff --git a/appimagetoolnoglib.c b/appimagetoolnoglib.c
	index 6e751e9..8290adf 100644
	--- a/appimagetoolnoglib.c
	+++ b/appimagetoolnoglib.c
	@@ -270,7 +270,7 @@ int main (int argc, char **argv)
	       // printf("%s", data);
	       fwrite(data, size, 1, fpdst);
	 
	-      if(ftruncate(fileno(fpdst), 128*1024) != 0) {
	+      if(ftruncate(fileno(fpdst), size) != 0) {
	           die("Not able to write padding to destination file, aborting");
	       }
	       
(putting the squashfs right after the runtime, as the [runtime seems to expect?](https://github.com/probonopd/AppImageKit/blob/appimagetool/master/runtime.c#L171)

**Comment by probonopd** [comment]
Correct, the runtime calculates the size of the ELF runtime and assumes that the squashfs image starts right after it. This is done so that the size of the runtime can dynamically change. You can print out the offset of an AppImage by running it with the `--appimage-offset` argument.

**Comment by probonopd** [comment]
> I'm confused. Does CI test i386? The code path I changed in elf.c is only ever executed against 32-bit ELFs, right?

No, but the PR apparently broke x86_68. Or did I make a mistake in testing it? Can you build a 64-bit one without your patch, then apply your patch and build another 64-bit one? Then run both AppImages to see if they are both working? (I con't do it right now but will do once I have some time.)

**Comment by probonopd** [comment]
> I never did get appimagetool working properly on my CentOS 5 i386 environment

Interesting. I have never tried to compile it on CentOS 5 so far but am interested in making this happen. Same for i386/i686.

**Comment by mkrautz** [comment]
> Can you build a 64-bit one without your patch, then apply your patch and build another 64-bit one? Then run both AppImages to see if they are both working? (I con't do it right now but will do once I have some time.)

Can't do it at the moment. Maybe tonight...

**Comment by probonopd** [comment]
> To prevent this from happening again, we should "eat our own dogfood" and use appimagetool.AppImage in the CI process to build e.g, appimaged.AppImage; so that errors like this would block the Travis CI build job and not go unnoticed.

Implemented in 4dc7953
